src/main/java/de/tum/in/www1/artemis/repository/ProgrammingSubmissionRepository.java

```diff
@@ -45,9 +56,45 @@ default ProgrammingSubmission findFirstByParticipationIdAndCommitHashOrderByIdDe
         return findByParticipationIdAndCommitHashOrderByIdDescWithFeedbacksAndTeamStudents(participationId, commitHash).stream().findFirst().orElse(null);
     }
 
-    @EntityGraph(type = LOAD, attributePaths = "results")
     Optional<ProgrammingSubmission> findFirstByParticipationIdOrderBySubmissionDateDesc(long participationId);
 
+    @EntityGraph(type = LOAD, attributePaths = { "results" })
+    Optional<ProgrammingSubmission> findProgrammingSubmissionWithResultsById(long programmingSubmissionId);
+
+    /**
+     * Finds the first programming submission by participation ID, including its results, ordered by submission date in descending order. To avoid in-memory paging by retrieving
+     * the first submission directly from the database.
+     *
+     * @param programmingSubmissionId the ID of the participation to find the submission for
+     * @return an {@code Optional} containing the first {@code ProgrammingSubmission} with results, ordered by submission date in descending order,
+     *         or an empty {@code Optional} if no submission is found
+     */
+    default Optional<ProgrammingSubmission> findFirstByParticipationIdWithResultsOrderBySubmissionDateDesc(long programmingSubmissionId) {
+        var programmingSubmissionOptional = findFirstByParticipationIdOrderBySubmissionDateDesc(programmingSubmissionId);
+        if (programmingSubmissionOptional.isEmpty()) {
+            return Optional.empty();
+        }
+        var id = programmingSubmissionOptional.get().getId();
+        return findProgrammingSubmissionWithResultsById(id);
+    }
+
+    @Query("""
+            SELECT new de.tum.in.www1.artemis.repository.ProgrammingSubmissionAndSubmissionDate(s, s.submissionDate)
```

Comment:
Why do you select the submission date as well only to later discard it and only use the id of the submission?